### PR TITLE
Make thread safe statistics accounting in TStorageFactoryFile

### PR DESCRIPTION
**Pull request by Dr15Jones**
#### PR description:

- Avoid race condition when setting up the storage accounting counters. 
- Added README to document classes and accounting details.

#### PR validation:

Code compiles.

### CODE REVIEW

**Comment by Dr15Jones** [comment]
please test

**Comment by Dr15Jones** [comment]
Please test

File: IOPool/TFileAdaptor/README.md [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+## Introduction
+The package contains a plugin to ROOT to allow a custom `TFile` to be generated by calls to `TFile::Open`. The custom class internally uses the `Storage` object from Utilities/StorageFactory.
+
+## TStorageFactorySystem
+A ROOT plugin which allows the use of `TStorageFactoryFile` when `TFile::Open` is called.
+
+## TStorageFactoryFile
+A specialization of `TFile` which uses a `edm::storage::Storage` instance internally.
+### storage accounting
+The `read` measurement statistics only encompasses a single buffer read. `readActual` encompasses both a single buffer read and a vector read. Only `readAsync` measures statistics for reading asynchronously (prefetching?).
```
Comment (makortel) [inline IOPool/TFileAdaptor/README.md:12]: It seems to me the situation is more complicated
https://github.com/cms-sw/cmssw/blob/49b3df1092701862f4113e36e7c0c35b958ad1d7/IOPool/TFileAdaptor/src/TStorageFactoryFile.cc#L243-L250

IIUC something like
* `readActual` is not just sum of `read` and `readv` (although apparently we don't record `tstoragefile-readv`)
* `tstoragefile-read` contains also other stuff (like reading from an already-cached buffer) than "actual read)

On the other hand, I'd expect what I wrote above to be incomplete as well, so I'm not sure if it is worth of using this PR to polish the text given that we expect updates.

Comment (Dr15Jones) [reply IOPool/TFileAdaptor/README.md:12]: Is there a change you want to be made?

Comment (makortel) [reply IOPool/TFileAdaptor/README.md:12]: I'd like to eventually see an accurate description of the various time measurements. So maybe it is better to not spend time in expanding the description right now.

**Comment by makortel** [comment]
+core

**Comment by rappoccio** [comment]
+1

**Pull request merged**